Support for Compiler Explorer

AJMansfield opened the pull request:
### Summary
<img width="1853" height="2013" alt="image" src="https://github.com/user-attachments/assets/fc2b64cd-0010-4bba-a5ba-c17eff479d99" />

This PR adds a `--json` option to mpy-tool, in order to generate Compiler-Explorer-compatible JSON annotation information for the bytecode disassembly. Some of this information might be theoretically possible to parse out from the text itself, but not all of it is, e.g. disambiguating child references with non-unique simple names.

- https://github.com/micropython/micropython/pull/18381
- https://github.com/compiler-explorer/misc-builder/pull/117
- https://github.com/compiler-explorer/compiler-workflows/pull/43
- https://github.com/compiler-explorer/infra/pull/1888
- https://github.com/compiler-explorer/compiler-explorer/pull/8256


### Testing
I've tested these changes against a locally-hosted version of Compiler Explorer, and can confirm it functions appropriately!

dpgeorge commented on tools/mpy-tool.py:
```diff
@@ -24,6 +24,10 @@
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 # THE SOFTWARE.
 
+from __future__ import annotations
```
Is this necessary?

dpgeorge commented on tools/mpy-tool.py:
```diff
@@ -1833,6 +1946,124 @@ def extract_segments(compiled_modules, basename, kinds_arg):
+
+
+old_print = print
+print = line_track_meta_print
```
Is there a better (?) way to do this, rather than having globals, have them as class variables of `JsonPrinter` and override the built-in `print` only if needed (eg when `JsonPrinter` is initialized)?

dpgeorge commented on tools/mpy-tool.py:
```diff
@@ -1833,6 +1946,124 @@ def extract_segments(compiled_modules, basename, kinds_arg):
+global_output_line_number: int = 0
+
+
+def line_track_meta_print(*a, meta: PartialAsmResultLine = {}, **k):
```
Can this be a `@staticmethod` of `JsonPrinter` to keep everything self contained?

AJMansfield committed:
tools/mpy-tool.py: Add Compiler Explorer JSON output.

This commit adds a `--json` option to `mpy-tool.py`, in order to generate
Compiler-Explorer-compatible JSON annotation information for the bytecode
disassembly. Some of this information might be theoretically possible to
parse out from the text itself, but not all of it is, e.g. disambiguating
child references with non-unique simple names.

Signed-off-by: Anson Mansfield <amansfield@mantaro.com>

dpgeorge commented on tools/mpy-tool.py:
```diff
@@ -985,6 +1005,75 @@ def freeze_raw_code(self, prelude_ptr=None, type_sig=0):
         raw_code_content += 4 * 4
 
+    @staticmethod
+    def decode_lineinfo(line_info: memoryview) -> "tuple[int, int, memoryview]":
```
I'm confused why the type annotations are within a string?  Doesn't that render them unusable?

AJMansfield replied:
Oh, I think I just quoted it on autopilot when I removed the `from __future__ import annotations` line --- putting annotations in quotes is a pure formal equivalent, with the evaluated value of the annotation in CPython just being a string with the annotation body either way.

dpgeorge replied:
OK, thanks for the explanation.

I understand that type hinting is useful, so I guess it's OK to keep these in here (even though the rest of this file does not have annotations).

dpgeorge commented on tools/mpy-tool.py:
```diff
@@ -985,6 +1005,75 @@ def freeze_raw_code(self, prelude_ptr=None, type_sig=0):
+                output.write(sep)
+            startCol = output.tell() + 1
+            output.write(child.simple_name.str)
+            endCol = output.tell() + 1
```
`end_col`

dpgeorge reviewed (commented):
This approach with the `print` shim looks pretty good now, it's nice and self contained.

AJMansfield commented:
Also added one more thing, to have it pick the languageId in the output depending on what's being output. (The 'stderr' mode is the one that interprets ANSI color codes.)
